src/class.c

```diff
@@ -1634,22 +1661,27 @@ bool InterfaceDeclaration::isBaseOf(ClassDeclaration *cd, int *poffset)
     {
         BaseClass *b = cd->interfaces[j];
 
-        //printf("\tbase %s\n", b->base->toChars());
-        if (this == b->base)
+        //printf("\tX base %s\n", b->sym->toChars());
+        if (this == b->sym)
         {
             //printf("\tfound at offset %d\n", b->offset);
             if (poffset)
             {
                 *poffset = b->offset;
-                if (j && cd->isInterfaceDeclaration())
+                // The conversion is interface(cd) to interface(this).
+                if (cd->isInterfaceDeclaration())
```

Comment:
It's been some time since I investigated the assembly of virtual inheritence (we use it to some extend at work). IIRC it was a series of indirections, but no extensive search. That might actually be enabled by a bit more layout information than currently available in D.

I think there is an important difference, though: while C++ actually has to find the correct instance of the virtual base class (that might contain data members), this is not important in D, because there are only interfaces without data members. Any instance of the vtbl that contains the correct function pointers will do.